backends-velox/src/main/scala/org/apache/gluten/execution/ColumnarCollectTailExec.scala

```diff
@@ -0,0 +1,142 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gluten.execution
+
+import org.apache.gluten.backendsapi.BackendsApiManager
+import org.apache.gluten.columnarbatch.ColumnarBatches
+import org.apache.gluten.columnarbatch.VeloxColumnarBatches
+import org.apache.gluten.extension.columnar.transition.Convention
+import org.apache.gluten.iterator.Iterators
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.serializer.Serializer
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.catalyst.plans.physical.SinglePartition
+import org.apache.spark.sql.execution.{ShuffledColumnarBatchRDD, SparkPlan}
+import org.apache.spark.sql.execution.metric.{SQLMetric, SQLShuffleWriteMetricsReporter}
+import org.apache.spark.sql.metric.SQLColumnarShuffleReadMetricsReporter
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+import scala.collection.mutable
+
+case class ColumnarCollectTailExec(
+    limit: Int,
+    child: SparkPlan
+) extends ColumnarCollectTailBaseExec(limit, child) {
+
+  private def collectTailRows(
+      partitionIter: Iterator[ColumnarBatch],
+      limit: Int
+  ): Iterator[ColumnarBatch] = {
+
+    if (!partitionIter.hasNext || limit <= 0) {
+      return Iterator.empty
+    }
+
+    val tailQueue = new mutable.ListBuffer[ColumnarBatch]()
+    var totalRowsInTail = 0L
+
+    while (partitionIter.hasNext) {
+      val batch = partitionIter.next()
+      val batchRows = batch.numRows()
+      ColumnarBatches.retain(batch)
+      tailQueue += batch
+      totalRowsInTail += batchRows
+
+      while (totalRowsInTail > limit && tailQueue.nonEmpty) {
+        val front = tailQueue.remove(0)
+        val frontRows = front.numRows().toLong
+        val overflow = totalRowsInTail - limit
+
+        if (frontRows <= overflow) {
+          totalRowsInTail -= frontRows
+        } else {
+          val keep = frontRows - overflow
+          val partial = VeloxColumnarBatches.slice(front, overflow.toInt, keep.toInt)
+          tailQueue.prepend(partial)
+          totalRowsInTail -= overflow
+        }
+        front.close()
+      }
+    }
+    Iterators
+      .wrap(tailQueue.iterator)
+      .recyclePayload(batch => batch.close())
+      .create()
+  }
+
+  private lazy val writeMetrics =
+    SQLShuffleWriteMetricsReporter.createShuffleWriteMetrics(sparkContext)
+
+  private lazy val readMetrics =
+    SQLColumnarShuffleReadMetricsReporter.createShuffleReadMetrics(sparkContext)
+
+  private lazy val useSortBasedShuffle: Boolean =
+    BackendsApiManager.getSparkPlanExecApiInstance
+      .useSortBasedShuffle(outputPartitioning, child.output)
```

Comment:
Also here https://github.com/apache/incubator-gluten/pull/9074#discussion_r2055703829 
https://github.com/apache/incubator-gluten/pull/9139/files#diff-25e08fa25b94cc139cb036eac3dd5c958d0aeb7b8570ffa6b020bcae27191a7d

Reply:
Just move it to `ColumnarCollectTailExecBase` to let CH backend use. The code is same for CH backend.